README.md
# NAME

App::grops::prepro - groff grops prepro

# SYNOPSIS

    use App::grops::prepro;
    exit(App::grops::prepro->run);

# DESCRIPTION

App::grops::prepro is ...

# LICENSE

Copyright (C) KUBO, Koichi.

This library is free software; you can redistribute it and/or modify
it under the same terms as Perl itself.

# AUTHOR

KUBO, Koichi <k@obuk.org>


Changes
Revision history for Perl extension App-grops-prepro

{{$NEXT}}

    - original version



lib/App/grops/prepro.pm
package App::grops::prepro;
use 5.008001;
use strict;
use warnings;

our $VERSION = "0.01";

use feature qw/say/;
use open qw/:locale :std/;      # debug outputs

use Class::Accessor 'antlers';
use Encode;
use File::Basename;
use Getopt::Std;
use Data::Clone 'clone';

sub InPUA       { "F0000 FFFFF" }
sub InDNL       { "F0000 F0FFF" }
sub InESC       { "F1000 F1FFF" }

has prologue    => (is => 'rw');
has re_req      => (is => 'rw');
has re_dnl      => (is => 'rw');
has re_esc      => (is => 'rw');
has pua         => (is => 'rw');
has pua_saving  => (is => 'rw');
has end         => (is => 'rw');
has unget       => (is => 'rw');
has tee         => (is => 'rw');
has debug       => (is => 'rw');

sub init {
  my ($self) = @_;

  $self->debug(0)  unless defined $self->debug;
  $self->pua({})   unless defined $self->pua;

  $self->unget([]) unless defined $self->unget;
  $self->end([])   unless defined $self->end;

  unless (defined $self->re_req) {
    $self->re_req(
      qr/ ^[.'] /x
    );
  }

  unless (defined $self->re_dnl) {
    $self->re_dnl(
      qr/ \\ (?: c.* )? $ /x
    );
  }

  unless (defined $self->re_esc) {
    $self->re_esc(
      qr/ \\ (?:
            \! .* $
          | \" .* $
          | \# .* $
          | \$ (?: (?&name) | \* | 0 | \@ | \^ )
          | \%
          | \&
          | \'
          | \)
          | \* (?&name)
          | \,
          | \-
          | \.
          | \/
          | 0
          | :
          | \? .*? \\\?
          | A (?&ident)
          | a
          | B '[^']*'
          | b '[^']*'
          | c .* $
          | C '[^']*'
          | d
          | D '[^']*'
          | e
          | E
          | f (?&name)
          | F (?&name)
          | g (?&name)
          | H '[^']*'
          | h (?: '[^']*' | \[ [^\]]* \] )
          | k (?&name)
          | l '[^']*'
          | L '[^']*'
          | m (?&name)
          | M (?&name)
          | n (?&name)
          | o '[^']*'
          | p
          | R '[^']*'
          | r
          | $
          | S '[^']*'
          | s (?&size)
          | \x{20}
          | u
          | t
          | v '[^']*'
          | w '[^']*'
          | x '[^']*'
          | X '[^']*'
          | Y (?&name)
          | Z (?: '[^']*' )
          | z (?&glyph)
          | \\
          | \^
          | \_
          | \'
          | \`
          | \{
          | \|
          | \}
          | \~
          | \[ [^\]]* \]
          | \( ..
          )
          (?(DEFINE)
            (?<ident> [^\s\x{00}\x{01}\x{0B}\x{0D}-\x{1F}\x{80}-\x{9F}]+ )
            (?<name>  [^\[\(] | \( .. | \[ [^\]]* \] )
            (?<glyph> \\ (?&name) | [^\\] )
            (?<size> [+\-]?\d | \([+\-]?\d\d )
          )
        /mx
      );
  }

  $self;
}


sub run {
  my $class = shift;
  $class->new(@_)->process();
}


sub new {
  my $class = shift;
  $class->SUPER::new(@_)->init();
}


sub process {
  my ($self) = @_;

  my @prepro;
  my $troff = join '', grep defined, $ENV{GROFF_COMMAND_PREFIX}, 'troff';
  push @prepro, shift @ARGV while @ARGV && $ARGV[0] ne $troff;
  my @troff = @ARGV;
  shift @ARGV;
  getopts('abcivzCERUw:W:d:f:m:n:o:r:T:F:I:M:', \ my %troff);

  # show subprograms version
  say join ' ', basename($0), "version", $VERSION
    if grep defined && $_ eq -v, @prepro;

  if (@troff) {
    local $ENV{PATH} = join ':', grep defined, $ENV{GROFF_BIN_PATH}, $ENV{PATH};
    open STDOUT, "|-", @troff or die usage();
  } elsif (@troff) {
    unshift @ARGV, @troff;
    @troff = ();
  }

  my $tee = $self->tee;
  open STDOUT, "|-", "tee", $tee or die "can't open $tee: $!" if $tee;
  $self->prepro_main if !grep $_ eq -v, @prepro;
  close STDOUT;
  return ($? >> 8);
}


sub usage {
  my @usage = (@_, "\n") if @_;
  push @usage, "usage: ", basename($0), " grops-opts troff troff-opts", "\n";
  @usage;
}

sub prepro_main {
  my ($self) = @_;
  if (my $prologue = $self->prologue) {
    $self->puts($prologue);
    $self->puts(".lf " . ($. // 0));
  }
  my $oldpua = $self->pua();
  do {
    $self->pua(clone($oldpua)) if $self->pua_saving;
  } while ($self->prepro_line);
  $self->pua_stats() if $self->debug & 32;
}


sub prepro_line {
  my ($self) = @_;

  my $req = $self->re_req;
  while (defined $self->gets()) {
    if ($self->debug & 1) {
      local $_ = $_;
      s/\p{InPUA}/$self->pua->{$&}/eg;
      s/\n/"\n" . (" " x 8)/eg;
      printf STDERR "%6d%s %s\n", $., @{$self->unget} == 0 ? ':' : ';', $_;
    }
    if (@{$self->end}) {
      my $end = $self->end->[-1];
      pop @{$self->end} if /$req\s*$end$/;
    }
    if (/$req\s*(de\S*|am\S*|ig)\s*(.*)/) {
      my ($f, $l) = ($1, $2);
      push @{$self->end}, "\\.";
      my @list = split /\s+/, $l;
      shift @list if $f =~ /^(de|am)/;
      if (defined $list[-1]) {
        my $end = join '', map { s/[\\.'"\[\](){}]/\\$&/; $_ } split //, $list[-1];
        push @{$self->end}, "(?:" . join('|', pop(@{$self->end}), $end) . ")";
      }
    } else {
      $self->prepro;
    }
    $self->puts;

    return 1 unless @{$self->unget};
  }

  undef;
}


sub prepro {
  my ($self) = @_;
}


sub pua_char {
  my ($self, $token, $cs) = @_;
  if (ref $cs eq 'CODE') {
    unless ($self->pua->{$cs}) {
      my ($free, $end) = map hex($_), split /\s+/, &$cs;
      $self->pua->{$cs} = { free => $free, start => $free, end => $end };
    }
    $token =~ s{\p{InPUA}}{$self->pua->{$&}}eg;
    unless (defined $self->pua->{token}{$token}) {
      if ($self->pua->{$cs}{free} > $self->pua->{$cs}{end}) {
        die "can't allocate pua($token, ",
          sprintf("%X..%X", $self->pua->{$cs}{start}, $self->pua->{$cs}{end}),
          "); try to run with pua_saving => 1\n";
      }
      my $code = pack "U", $self->pua->{$cs}{free}++;
      if ($self->debug & 16) {
        say STDERR "# _roffchar: \$token = $token, \$cs = $cs, \$code = ",
          sprintf("\\x{%X}", ord $code);
      }
      $self->pua->{$code} = $token;
      $self->pua->{token}{$token} = $code;
    }
    $self->pua->{token}{$token};
  } else {
    my $code = $cs;
    $self->pua->{$code} = $token;
    $self->pua->{token}{$token} = $code;
  }
}


sub pua_stats {
  my ($self) = @_;

  my %used;
  for (keys %{$self->pua}) {
    next unless ref $self->pua->{$_};
    next unless exists $self->pua->{$_}{free};
    $used{$self->pua->{$_}{start}} = {
      a => $self->pua->{$_}{free} - $self->pua->{$_}{start},
      b => $self->pua->{$_}{end} - $self->pua->{$_}{start},
    };
  }
  say STDERR '# pua stats: ', join ', ', map { sprintf("%X: %s", $_, "$used{$_}{a}/$used{$_}{b}") } sort keys %used;
}


sub puts {
  my $self = shift;
  if (@_) {
    $self->puts() for @_;
  } else {
    my @lines = split /\n/;
    @lines = "" unless @lines;
    for (@lines) {
      s{\p{InPUA}}{
        my $subst = $self->pua->{$&};
        $subst .= ".lf $.\n" if $subst =~ /\n$/;
        $subst;
      }eg;
      $self->putline;
    }
  }
}


sub putline {
  my ($self) = @_;
  say;
}


sub gets {
  my ($self) = @_;

  my $req = $self->re_req;

  return undef unless defined $self->getline();

  while (/\p{InDNL}$/) {
    my $line = $_;
    if (defined $self->gets()) {
      $line .= "\n" . $_;
    } else {
      $_ = $line;
      last;
    }
    $_ = $line;
  }

  $. = $1 - 1 if /$req\s*lf\s+(\d+)/;

  return $_;
}


sub getline {
  my ($self) = @_;

  my $esc = $self->re_esc;
  my $dnl = $self->re_dnl;

  if (defined ($_ = shift @{$self->unget})) {
    ;
  } elsif (defined ($_ = <>)) {
    chomp;
    1 while s{$esc}{
      my $e = $&;
      if ($e =~ /^$dnl/) {
        $self->pua_char($e, \&InDNL);
      } else {
        $self->pua_char($e, \&InESC)
      }
    }e;
  }
  $_;
}

1;
__END__

=encoding utf-8

=head1 NAME

App::grops::prepro - groff grops prepro

=head1 SYNOPSIS

    use App::grops::prepro;
    exit(App::grops::prepro->run);

=head1 DESCRIPTION

App::grops::prepro is ...

=head1 LICENSE

Copyright (C) KUBO, Koichi.

This library is free software; you can redistribute it and/or modify
it under the same terms as Perl itself.

=head1 AUTHOR

KUBO, Koichi E<lt>k@obuk.orgE<gt>

=cut



lib/App/grops/prepro/JA/CC.pm
package App::grops::prepro::JA::CC;
use 5.008001;
use strict;
use warnings;

our $VERSION = "0.01";

=encoding utf-8

=head1 NAME

App::grops::prepro::JA::CC - It's new $module

=head1 SYNOPSIS

    use App::grops::prepro::JA::CC;

=cut

use App::grops::prepro::JA::W3C_jlreq ':all';

use feature 'say';
use Exporter 'import';
our @ISA = qw(Exporter);
our @EXPORT_OK = ();

{
  my $package = __PACKAGE__ . "::";
  no strict 'refs';
  push @EXPORT_OK, grep /^(In|Is)/, keys %{$package};
}

our %EXPORT_TAGS = (all => \@EXPORT_OK);


# user-defined props (alphabetical order)
sub InEnding {
  return <<END;
+utf8::InClose_Punctuation
+utf8::InFinal_Punctuation
+InColon
+InFullStops
+InCommas
+InDividingPunctuationMarks
END
}

sub InEndingJ {
  return <<END;
+InEnding
&InJapaneseCharacters
END
}


sub InEndingW {
  return <<END;
+InEnding
&InWesternCharacters
END
}


sub InIVS {
  "E0100 E01EF";
}


sub InInsep {
  (my $u = <<END) =~ s/[#;].*//gm; $u;
+InInseparableCharacters
+002F; \/
END
}


sub InJapanese {
  return <<END;
+InJapaneseCharacters
-InSVS
-InIVS
-InPunctuations
END
}


sub InJapaneseCharacters {
  # see $Config{privlib}/unicore/Blocks.txt
  (my $u = <<END) =~ s/[#;].*//gm; $u;
3000 303F; CJK Symbols and Punctuation
3040 309F; Hiragana
30A0 30FF; Katakana
3190 319F; Kanbun
31C0 31EF; CJK Strokes
31F0 31FF; Katakana Phonetic Extensions
3200 32FF; Enclosed CJK Letters and Months
3300 33FF; CJK Compatibility
3400 4DBF; CJK Unified Ideographs Extension A
4DC0 4DFF; Yijing Hexagram Symbols
4E00 9FFF; CJK Unified Ideographs
F900 FAFF; CJK Compatibility Ideographs
FE00 FE0F; Variation Selectors
FF00 FFEF; Halfwidth and Fullwidth Forms
20000 2A6DF; CJK Unified Ideographs Extension B
2A700 2B73F; CJK Unified Ideographs Extension C
2B740 2B81F; CJK Unified Ideographs Extension D
2B820 2CEAF; CJK Unified Ideographs Extension E
2CEB0 2EBEF; CJK Unified Ideographs Extension F
2F800 2FA1F; CJK Compatibility Ideographs Supplement
E0100 E01EF; Variation Selectors Supplement
END
}


sub InMiddleDotsJ {
  return <<END;
+InMiddleDots
&InJapaneseCharacters
END
}


sub InPunctuations {
  return <<END;
+InStarting
+InEnding
+InHyphens
+InMiddleDots
+utf8::InPunctuation
+utf8::InSymbol
END
}


sub InStarting {
  return <<END;
+utf8::InOpen_Punctuation
+utf8::InInitial_Punctuation
END
}


sub InStartingJ {
  return <<END;
+InStarting
&InJapaneseCharacters
END
}


sub InStartingW {
  return <<END;
+InStarting
&InWesternCharacters
END
}


sub InSVS {
  "FE00 FE0F";
}


sub InUnitSymbolsSimple {
  (my $u = <<END) =~ s/[#;].*//gm; $u;
+InUnitSymbols
-0020;	 	SPACE	quarter em width
-0028;	( 	LEFT PARENTHESIS
-0029;	) 	RIGHT PARENTHESIS
-002F;	\/ 	SOLIDUS one third em width, half-width or proportional
END
}


=begin comment

\p {InUSPC} is a user-defined property that matches spaces in the man
page source code.  This property also includes spaces created by
roff's escape \/, \^, \|, \h'nnn', etc., but not \n.  \n is used to
connect multiple continuation lines.

=end comment

=cut

sub InUSPC {
  return <<END;
+utf8::InSpace
-000A
+InSPC
END
}


sub InWestern {
  return <<END;
+InWesternCharacters
-InJapaneseCharacters
END
}

1;


lib/App/grops/prepro/JA.pm
package App::grops::prepro::JA;
use 5.008001;
use strict;
use warnings;

our $VERSION = "0.01";

use feature qw/say/;
use parent 'App::grops::prepro';

use utf8;
use App::grops::prepro::JA::CC ':all';
use Class::Accessor 'antlers';

sub InPUA       { "F0000 FFFFF" }
sub InPSPC      { "F0000 F000F" } # <-> InUSPC
sub InDNL       { "F0010 F01FF" }
sub InSPC       { "F0200 F0FFF" }
sub InESC       { "F1000 F1FFF" }
sub InFCD       { "F2000 F2FFF" }
sub InNUM       { "F3000 F3FFF" }

has emsp        => (is => 'rw');
has hemsp       => (is => 'rw');
has qemsp       => (is => 'rw');
has zwsp        => (is => 'rw');
has nrsp        => (is => 'rw');
has wdsp        => (is => 'rw');
has cr          => (is => 'rw');

has re_num      => (is => 'rw');
has re_mode     => (is => 'rw');
has re_spc      => (is => 'rw');
has re_esc      => (is => 'rw');
has fc          => (is => 'rw');

sub m_punct     { (1 << 0) }
sub m_zwsp      { (1 << 1) }
sub m_wdsp      { (1 << 2) }
sub m_nrsp      { (1 << 4) }
sub m_cr        { (1 << 5) }

sub init {
  my ($self) = @_;

  eval { require Unicode::Normalize };

  #$self->debug(16);

  unless (defined $self->mode) {
    $self->mode(m_punct | m_zwsp | m_wdsp | m_nrsp);
  }

  unless (defined $self->re_mode) {
    $self->re_mode(qr/\\"\s*pp-ja/);
  }

  unless (defined $self->prologue()) {
    $self->prologue(<<'END');
.nr pp:debug 0
.ds pp:color orange
.de pp:cr
\c
.  if (\\n[pp:debug]) \{\
.    ie '\\*[pp:color]'' .nop \(CR\c
.    el .nop \m[\\*[pp:color]]\(CR\m[]\c
.  \}
..
.de pp:sp
\c
.  nr \\$0.ss  \\n[.ss]
.  nr \\$0.sss \\n[.sss]
.  ss
.  ie \\n[.$] .ss \\$*
.  el .ss \\n[\\$0-width]
.  nr \\$0-w (\\n[.ss]*100)u
.  if (\\n[pp:debug])&(\\n[.ss]) \{\
.    ie '\\*[pp:color]'' .nop \Z'\D'l 0 +0.1'\D'l +\\n[\\$0-w]u 0''\c
.    el .nop \m[\\*[pp:color]]\Z'\D'l 0 +0.1'\D'l +\\n[\\$0-w]u 0''\m[]\c
.  \}
.  nop \& \c
.  if (\\n[pp:debug])&(\\n[.ss]) \{\
.    ie '\\*[pp:color]'' .nop \Z'\D'l 0 +0.1'\D'l -\\n[\\$0-w]u 0''\c
.    el .nop \m[\\*[pp:color]]\Z'\D'l 0 +0.1'\D'l -\\n[\\$0-w]u 0''\m[]\c
.  \}
.  ss
.  ss \\n[\\$0.ss] \\n[\\$0.sss]
..
.nr pp:emsp-width  (\n[.ss] * 320 / 100)
.nr pp:hemsp-width (\n[.ss] * 160 / 100)
.nr pp:qemsp-width (\n[.ss] *  80 / 100)
.nr pp:nrsp-width  (\n[.ss] *  20 / 100)
.nr pp:wdsp-width  (\n[.ss])
.nr pp:zwsp-width  0
.
.als pp:emsp  pp:sp
.als pp:hemsp pp:sp
.als pp:qemsp pp:sp
.als pp:nrsp  pp:sp
.als pp:wdsp  pp:sp
.als pp:zwsp  pp:sp
END
  }

  $self->emsp("\\*[pp:emsp ]")   unless defined $self->emsp;
  $self->hemsp("\\*[pp:hemsp ]") unless defined $self->hemsp;
  $self->qemsp("\\*[pp:qemsp ]") unless defined $self->qemsp;
  $self->wdsp("\\*[pp:wdsp ]")   unless defined $self->wdsp;
  $self->nrsp("\\*[pp:nrsp ]")   unless defined $self->nrsp;
  $self->zwsp("\\*[pp:zwsp ]")   unless defined $self->zwsp;

  $self->cr("\\*[pp:cr ]")       unless defined $self->cr;

  unless (defined $self->re_spc) {
    $self->re_spc(
      qr/ \\ (?:
            \/                    # Ligatures and Kerning
          | \,
          | \&
          | \)

          | v (?: '[^']*' | \[ [^\]]* \] ) # Page Motions
          | r
          | u
          | d
          | h (?: '[^']*' | \[ [^\]]* \] )
          | \x{20}
          | \~
          | \|
          | \^
          | 0
          | w '[^']*'
          | k (?&name)
          | o '[^']*'
          | z (?&glyph)
          | Z '[^']*'
          )
          (?(DEFINE)
            (?<ident> [^\s\x{00}\x{01}\x{0B}\x{0D}-\x{1F}\x{80}-\x{9F}]+ )
            (?<name>  [^\[\(] | \( .. | \[ [^\]]* \] )
            (?<glyph> \\ (?&name) | [^\\] )
            (?<size> [+\-]?\d | \([+\-]?\d\d )
          )
        /mx
      );
  }

  unless (defined $self->re_num) {
    $self->re_num(
      qr/ (?&pre)? (?&num) (?&post)?
          (?(DEFINE)
            (?<pre> \p{InPrefixedAbbreviations}+ )
            (?<num> \d+ (?: (?:[.,]\p{InUSPC}?) \d+ )* )
            (?<post> (?: \p{InPostfixedAbbreviations} | \p{InUnitSymbolsSimple} )+
            | \p{InUSPC}* [(\[] [^\)\]]+ [)\]] )
          )
        /x
      );
  }

  $self->SUPER::init;

  for (qw/ IsEmSp IsHEmSp IsQEmSp IsWdSp IsNrSp IsZwSp /) {
    /^Is(.*)/;
    my $sp = lc($1);
    my $_sp = "_$sp";
    my $c = $self->pua_char($_, \&InPSPC);
    $self->pua_char($self->$sp, $c);
    eval sprintf 'sub %s { "%X" }', $_, ord $c;
    eval sprintf 'sub %s { "\\x{%X}" }', $_sp, ord $c;
  }

  $self;
}


sub prepro {
  my ($self) = @_;

  my $req  = $self->re_req;
  my $mode = $self->re_mode;

  if ($mode) {
    if (my ($e) = /$req\s*(\p{InESC})/) {
      if ($self->pua->{$e} =~ /$mode(?:\s+(\d+))?/) {
        $self->mode($1);
      }
    }
  }
  if (/$req\s*fc(?:\s+(.)(.)?)?$/) {
    if (defined $1) {
      $self->fc(sprintf "\\x{%X}", ord $1);
    } else {
      $self->fc(undef);
    }
  } elsif (/$req/) {

    if (/$req\s*Sh\s+(名前|名称)\b/) {
      $_ = join "\n",
        ".ds section-name $1",
        ".ds doc-section-name $1",
        ".lf $.",
        $_;
    }

  } else {

    my $m = $self->mode;

    if (my $delim = $self->fc) {
      s/${delim}.*${delim}/$self->pua_char($&, \&InFCD)/e;
    }

    if ($m & m_zwsp) {
      s{\p{InJapanese}{2,}|[\p{InSVS}\p{InIVS}]\p{InJapanese}}{
        join $self->_zwsp(), split //, $&;
      }eg;

      s{(\p{InJapanese})(\p{InDNL}\n)(\p{InJapanese})}{
        my ($j1, $c, $j2) = ($1, $2, $3);
        ($j1 =~ /\p{InStarting}/ || $j2 =~ /\p{InEnding}/)?
          $j1.$c.$j2 : $j1.$self->_zwsp().$c.$j2;
      }eg;
    }

    if ($m & m_wdsp) {
      s{(\p{InJapanese}[\p{InSVS}\p{InIVS}]?)(\p{InWestern})}{do {
        my ($x, $p) = ($1, $2);
        ($p =~ /^\p{InEnding}/)? $x.$p : $x.$self->_wdsp().$p;
      }}eg;

      s{(\p{InWestern})(\p{InJapanese})}{do {
        my ($p, $x) = ($1, $2);
        ($p =~ /\p{InStarting}$/)? $p.$x : $p.$self->_wdsp().$x;
      }}eg;
    }

    if ($m & m_nrsp) {
      if (my $number = $self->re_num) {
        s{$number}{$self->pua_char($&, \&InNUM)}egx;
        s{(\p{InJapanese}[\p{InSVS}\p{InIVS}]?)\p{IsWdSp}*(\p{InNUM})}{
          $1.$self->_nrsp.$2;
        }egx;
        s{(\p{InNUM})\p{IsWdSp}*(\p{InJapanese})}{
          $1.$self->_nrsp.$2;
        }egx;
      }
    }

    if ($m & m_punct) {
      # 3.1.2
      s/\p{InPSPC}*(\p{InStartingJ})/
        $self->_hemsp().$1/eg;
      s/(\p{InEndingJ})\p{InPSPC}*/
        $1.$self->_hemsp()/eg;
      s/\p{InPSPC}*(\p{InMiddleDotsJ}+)\p{InPSPC}*/
        $self->_qemsp().$1.$self->_qemsp()/eg;

      s/(\p{InJapanese}[\p{InSVS}\p{InIVS}]?)(\p{InStartingW})/
        $1.$self->_wdsp().$2/eg;
      s/(\p{InEndingW})(\p{InJapanese})/
        $1.$self->_wdsp().$2/eg;

      # 3.1.4
      s/\p{InPSPC}+(\p{InEnding})/$1/g;
      s/(\p{InStarting})\p{InPSPC}+/$1/g;

    }

    # remove \p{InPSPC} in quotes
    s/([\`\'\"])\p{InPSPC}+(.*?)\p{InPSPC}+([\'\"])/do {
      $1.$2.$3;
    }/eg;

    # remove \p{InPSPC} around \p{InInsep} characters
    s/\p{InPSPC}*(\p{InInsep}+)\p{InPSPC}*/$1/g;

    # to prefer input, remove \p{InPSPC} adjacent to \p{InUSPC}.
    s/\p{InPSPC}+(\p{InUSPC})/$1/g;
    s/(\p{InUSPC})\p{InPSPC}+/$1/g;

    s/^\p{InPSPC}+//sg;
    s/\p{InPSPC}+$//sg;

    if ($m & m_cr) {
      s/$/$self->cr/meg;
    }

  }
}

sub mode {
  my $self = shift;
  $self->{mode} //= [];
  if (@_) {
    if (defined $_[0]) {
      push @{$self->{mode}}, $_[0];
    } else {
      pop @{$self->{mode}};
    }
  }
  ${$self->{mode}}[-1];
}


sub gets {
  my ($self) = @_;

  return undef unless defined $self->SUPER::gets();

  my $m = $self->mode;

  my $req = $self->re_req;
  my $ec = $self->pua_char("\\c", \&InDNL);

  while (defined && !/$req/ && !($m & m_cr) && /\p{InJapaneseCharacters}$/) {
    my $line = $_;
    if (defined $self->SUPER::gets()) {
      if (!($m & m_cr) && /^\p{InJapaneseCharacters}/) {
        $line .= $ec . "\n" . $_;
        $_ = $line;
      } else {
        push @{$self->{unget}}, $_;
        $_ = $line;
        last;
      }
    } else {
      $_ = $line;
      last;
    }
  }

  return $_;

}

sub putline {
  my ($self) = @_;

  s{([^[:ascii:]\p{InSVS}\p{InIVS}])([\p{InSVS}\p{InIVS}])?}{
    my ($u, $vs) = ($1, $2);
    if (__PACKAGE__->can('NFD')) {
      $u = NFD($u);
    }
    my @u = unpack "U*", $u;
    push @u, unpack "U*", $vs if $vs;
    sprintf "\\[u".join('_', ("%04X") x @u)."]", @u;
  }eg;

  say;
}


sub getline {
  my ($self) = @_;

  my $esc = $self->re_esc;
  $esc = qr/$esc/;

  my $dnl = $self->re_dnl;
  my $spc = $self->re_spc;

  if (defined ($_ = shift @{$self->{unget}})) {
    ;
  } elsif (defined ($_ = <>)) {
    my $newline = chomp;

    1 while s{$esc}{
      my $e = $&;
      if ($e =~ /^$spc$/) {
        $self->pua_char($e, \&InSPC);
      } elsif ($e =~ /^$dnl$/) {
        $self->pua_char($e, \&InDNL);
      } elsif ($e =~ /^\\\[u([0-9A-F_]+)\]/) {
        my @u = map { pack "U", hex } split '_', $1;
        my $vs;
        if ($u[-1] =~ /[\p{InSVS}\p{InIVS}]/) {
          $vs = pop @u;
        }
        my $u = join '', @u;
        if (__PACKAGE__->can('NFC')) {
          $u = NFC($u);
        }
        $u .= $vs if $vs;
        $u;
      } else {
        $self->pua_char($e, \&InESC)
      }
    }e;

    $_ .= $self->pua_char("\\c", \&InDNL) unless $newline;

  }

  $_;
}

1;
__END__

=encoding utf-8

=head1 NAME

App::grops::prepro::JA - It's new $module

=head1 SYNOPSIS

    use App::grops::prepro::JA;
    exit(App::grops::prepro::JA->run);

=head1 DESCRIPTION

App::grops::prepro::JA is ...

=head1 LICENSE

Copyright (C) KUBO, Koichi.

This library is free software; you can redistribute it and/or modify
it under the same terms as Perl itself.

=head1 AUTHOR

KUBO, Koichi E<lt>k@obuk.orgE<gt>

=cut

